website/project/model.py

```diff
@@ -2829,188 +2910,302 @@ def to_json(self):
+    def _notify_non_authorizer(self, user):
+        pass
 
-    UNAPPROVED = 'unapproved'
-    ACTIVE = 'active'
-    CANCELLED = 'cancelled'
-    COMPLETED = 'completed'
+    def ask(self, group):
+        for contrib in group:
+            if contrib._id in self.approval_state:
+                self._notify_authorizer(contrib)
+            else:
+                self._notify_non_authorizer(contrib)
 
-    _id = fields.StringField(primary=True, default=lambda: str(ObjectId()))
-    initiation_date = fields.DateTimeField(auto_now_add=datetime.datetime.utcnow)
-    initiated_by = fields.ForeignField('user', backref='embargoed')
-    end_date = fields.DateTimeField()
-    # Expanded: Dictionary field mapping admin IDs their approval status and relevant tokens:
+class EmailApprovableSanction(Sanction):
+
+    AUTHORIZER_NOTIFY_EMAIL_TEMPLATE = None
+    NON_AUTHORIZER_NOTIFY_EMAIL_TEMPLATE = None
+
+    VIEW_URL_TEMPLATE = ''
+    APPROVE_URL_TEMPLATE = ''
+    REJECT_URL_TEMPLATE = ''
+
+    # Store a persistant copy of urls for use when needed outside of a request context.
+    # This field gets automagically updated whenever models approval_state is modified
+    # and the model is saved
     # {
-    #   'b3k97': {
-    #     'has_approved': False,
-    #     'approval_token': 'Pew7wj1Puf7DENUPFPnXSwa1rf3xPN',
-    #     'disapproval_token': 'TwozClTFOic2PYxHDStby94bCQMwJy'}
+    #   'abcde': {
+    #     'approve': [APPROVAL_URL],
+    #     'reject': [REJECT_URL],
+    #   }
     # }
-    approval_state = fields.DictionaryField()
-    # One of 'unapproved', 'active', 'cancelled', or 'completed
-    state = fields.StringField(default='unapproved', validate=validate_embargo_state)
-    for_existing_registration = fields.BooleanField(default=False)
+    stashed_urls = fields.DictionaryField()
 
-    def __repr__(self):
-        parent_registration = Node.find_one(Q('embargo', 'eq', self))
-        return ('<Embargo(parent_registration={0}, initiated_by={1}, '
-                'end_date={2}) with _id {3}>').format(
-            parent_registration,
+    @staticmethod
+    def _format_or_empty(template, context):
+        if context:
+            return template.format(**context)
+        return ''
+
+    def _view_url(self, user_id):
+        return self._format_or_empty(self.VIEW_URL_TEMPLATE, self._view_url_context(user_id))
+
+    def _view_url_context(self, user_id):
+        return None
+
+    def _approval_url(self, user_id):
+        return self._format_or_empty(self.APPROVE_URL_TEMPLATE, self._approval_url_context(user_id))
+
+    def _approval_url_context(self, user_id):
+        return None
+
+    def _rejection_url(self, user_id):
+        return self._format_or_empty(self.REJECT_URL_TEMPLATE, self._rejection_url_context(user_id))
+
+    def _rejection_url_context(self, user_id):
+        return None
+
+    def _send_approval_request_email(self, user, template, context):
+        mails.send_mail(
             self.initiated_by,
```

Comment:
This is currently causing an error when emails are sent. I believe this is supposed to be `self.initiated_by.username`. Please fix and write a regression test.